### ISSUE TITLE
Podman 1.7.0 incompatible with mount_program using overlay2 driver?

### ISSUE THREAD

**Issue by wwilson** [containers/libpod#4888]
Hi All,

We were successfully using fuse-overlayfs with overlay2 driver on podman 1.6.2.

Following upgrade to 1.7.0, the output of 'podman info' no longer shows any mount_program option. We noticed the change to the structure of storage.conf, and moved the setting inside [storage.options.overlay] block. Tried again and still didn't work. We also tried making a [storage.options.overlay2] block, and no luck there either.

Is there a recommendation for what we should try next?

**Comment by mheon** [comment]
@rhatdan @nalind Did we change the format of `storage.conf` recently?

**Comment by rhatdan** [comment]
Not in that area.  Are you sure you are looking at the correct storage.conf file.  The one in ~/.config/containers/storage.conf?

**Comment by wwilson** [comment]
That's the one we're looking at.

What is the recommended way to configure podman 1.7 to use fuse-overlayfs under overlay2? Does `podman info` still report whether it's using a custom mount program?

**Comment by rhatdan** [comment]
It should.  It certainly works for me.

Could you just remove the storage.conf file and let podman regenerate it?

```
$ rpm -q podman
podman-1.7.0-2.fc31.x86_64
$  podman --version
podman version 1.7.0
$  podman info | grep fuse
      Executable: /bin/fuse-overlayfs
      Package: fuse-overlayfs-0.7.2-2.fc31.x86_64
        fusermount3 version: 3.6.2
        fuse-overlayfs: version 0.7.2
```

**Comment by wwilson** [comment]
Hmm... I've tried running podman with --log-level=debug, and that didn't get me anything, but fuse-overlayfs isn't present at all in the output of `podman info` in 1.7.0.

I noticed I have a relatively old version of fuse-overlayfs. Is it possible there's some incompatibility which is causing a silent failure?

**Comment by rhatdan** [comment]
Perhaps.  @giuseppe Any ideas?

**Comment by wwilson** [comment]
Upgraded to fuse-overlayfs-0.7.3 and this is still not working.

FWIW, I am running on ubuntu-18.04

**Comment by rhatdan** [comment]
On Ubuntu the overlay driver works in the user homedir.  I believe, so you might not need fuse-overalay, maybe this is what is going on.  Do containers work for you?

**Comment by giuseppe** [comment]
No you can't still use overlay as the unprivileged user cannot create whiteout files and the storage driver fails trying to create them

**Comment by rhatdan** [comment]
Ok that is a fix that is going upstream.  Any idea what is going wrong with @wwilson box?

**Comment by wwilson** [comment]
Does that go for overlay2 as well?

If fuse-overlayfs is no longer needed for rootless containers with overlay2, then that will solve my problem... eventually.

**Comment by giuseppe** [comment]
@wwilson fuse-overlayfs should be used automatically when it is installed.

If not please make sure it is used as following:

```
[storage]
  driver = "overlay"
  [storage.options]
    mount_program = "/usr/bin/fuse-overlayfs"

```

**Comment by rhatdan** [comment]
Container engines do not differentiate between overlay and overlay2 (They are both overlay2)  No one uses the original overlay driver from Docker, so we just called our version of overlay2, overlay.

@wwilson No as @giuseppe points out, currently overlay is allowed to be mounted, as non root on Ubuntu, but you are still not allowed to create whiteout devices, since you don't have CAP_MKNOD.  Upstream is working on patches to both allow mounting of overlay as non root and then to allow creation of device nodes with major minor numbers of 0:0 to be created without CAP_MKNOD.  When this eventually bubbles up to distro kernels, we should be able to use kernel overlay in rootless mode.  There are other features of fuse-overlay, that might still make it preferable, but we shall see.

**Comment by wwilson** [comment]
I checked, and I'm getting the same behavior with buildah. So maybe a change in containers/storage?

Buildah info at 1.10.1 (note overlay2.mount_program):

> {
>     "host": {
>         "Distribution": {
>             "distribution": "ubuntu",
>             "version": "18.04"
>         },
>         "MemTotal": 67481702400,
>         "MenFree": 7973285888,
>         "SwapFree": 16381112320,
>         "SwapTotal": 16382947328,
>         "arch": "amd64",
>         "cpus": 8,
>         "hostname": "will-desktop",
>         "kernel": "4.15.0-74-generic",
>         "os": "linux",
>         "rootless": true,
>         "uptime": "167h 13m 6.84s (Approximately 6.96 days)"
>     },
>     "store": {
>         "ContainerStore": {
>             "number": 3
>         },
>         "GraphDriverName": "overlay",
>         "GraphOptions": [
>             "overlay2.mount_program=/home/will/src/fuse-overlayfs/fuse-overlayfs"
>         ],
>         "GraphRoot": "/home/will/.local/share/containers/storage",
>         "GraphStatus": {
>             "Backing Filesystem": "extfs",
>             "Native Overlay Diff": "false",
>             "Supports d_type": "true",
>             "Using metacopy": "false"
>         },
>         "ImageStore": {
>             "number": 15
>         },
>         "RunRoot": "/run/user/1001/containers"
>     }
> }

Buildah info at 1.12.0 (mount program missing):

> {
>     "host": {
>         "CgroupVersion": "v1",
>         "Distribution": {
>             "distribution": "ubuntu",
>             "version": "18.04"
>         },
>         "MemTotal": 67481702400,
>         "MenFree": 7987343360,
>         "OCIRuntime": "runc",
>         "SwapFree": 16381112320,
>         "SwapTotal": 16382947328,
>         "arch": "amd64",
>         "cpus": 8,
>         "hostname": "will-desktop",
>         "kernel": "4.15.0-74-generic",
>         "os": "linux",
>         "rootless": true,
>         "uptime": "167h 12m 19.29s (Approximately 6.96 days)"
>     },
>     "store": {
>         "ContainerStore": {
>             "number": 3
>         },
>         "GraphDriverName": "overlay",
>         "GraphOptions": null,
>         "GraphRoot": "/home/will/.local/share/containers/storage",
>         "GraphStatus": {
>             "Backing Filesystem": "extfs",
>             "Native Overlay Diff": "false",
>             "Supports d_type": "true",
>             "Using metacopy": "false"
>         },
>         "ImageStore": {
>             "number": 15
>         },
>         "RunRoot": "/run/user/1001/containers"
>     }
> }

Newly regenerated storage.conf (I set the mount program in both the locations that seem implied by the old format and by the new format):

> [storage]
>   driver = "overlay2"
>   runroot = "/run/user/1001/containers"
>   graphroot = "/home/will/.local/share/containers/storage"
>   [storage.options]
>     size = ""
>     remap-uids = ""
>     remap-gids = ""
>     ignore_chown_errors = ""
>     remap-user = ""
>     remap-group = ""
>     skip_mount_home = ""
>     mount_program = "/home/will/src/star/build_tools/fuse-overlayfs"
>     mountopt = ""
>     [storage.options.aufs]
>       mountopt = ""
>     [storage.options.btrfs]
>       min_space = ""
>       size = ""
>     [storage.options.thinpool]
>       autoextend_percent = ""
>       autoextend_threshold = ""
>       basesize = ""
>       blocksize = ""
>       directlvm_device = ""
>       directlvm_device_force = ""
>       fs = ""
>       log_level = ""
>       min_free_space = ""
>       mkfsarg = ""
>       mountopt = ""
>       size = ""
>       use_deferred_deletion = ""
>       use_deferred_removal = ""
>       xfs_nospace_max_retries = ""
>     [storage.options.overlay]
>       ignore_chown_errors = ""
>       mountopt = ""
>       mount_program = "/home/will/src/fuse-overlayfs/fuse-overlayfs"
>       size = ""
>       skip_mount_home = ""
>     [storage.options.vfs]
>       ignore_chown_errors = ""
>     [storage.options.zfs]
>       mountopt = ""
>       fsname = ""
>       size = ""

**Comment by rhatdan** [comment]
With that change does it work?

**Comment by wwilson** [comment]
Not sure what change you're referring to. The storage.conf is unchanged between the two invocations of "buildah info". It's just that after upgrading the mount_program option isn't reflected anywhere (and ditto for podman).

**Comment by rhatdan** [comment]
Does podman info show the fuse-overlay?

**Comment by wwilson** [comment]
It does at version 1.6.2, it does not at version 1.7.0.

**Comment by wwilson** [comment]
My guess that this could be due to containers/storage was it seemed unlikely that the same change was made to both buildah and podman, but presumably you guys have both vendored a newere containers/storage in the meantime?

**Comment by rhatdan** [comment]
I think the overlay2 might be confusing it.  Can you change your storage.conf to have just overlay not overlay2?

**Comment by mheon** [comment]
For clarity, containers/storage (and the programs that use it - Podman, Buildah, CRI-O) do not have an overlay2 driver. The original, deprecated overlay driver was removed, and overlay2 was renamed to overlay. I'm actually surprised that c/storage works at all with overlay2 set as the driver

**Comment by rhatdan** [comment]
It aliases them together.  But I found a bug.

https://github.com/containers/storage/pull/513

**Comment by rhatdan** [comment]
In my testing on my laptop setting overlay2 caused a similar issue, The above PR fixed it for me.

**Comment by wwilson** [comment]
Switching to "overlay" does appear to fix the problem. If "overlay" and "overlay2" in fact get us the exact same on-disk format, then I think we're all good here.

Thanks for the help all.

**Comment by phlogistonjohn** [comment]
I also had a fairly minimal storage.conf that no longer works:
```
[storage]
  graphroot = "/srv/bulk/users/jmulliga/containers/storage"
    
  [storage.options]
    mount_program = "/usr/bin/fuse-overlayfs"
```
And it now produces the following with 'podman images':
```
ERRO[0000] [graphdriver] prior storage driver overlay failed: kernel does not support overlay fs: 'overlay' is not supported over xfs at "/srv/bulk/users/jmulliga/containers/storage/overlay": backing file system is unsupported for this graph driver   
Error: Could not get runtime: kernel does not support overlay fs: 'overlay' is not supported over xfs at "/srv/bulk/users/jmulliga/containers/storage/overlay": backing file system is unsupported for this graph driver
```
Adding in just `driver = "overlay"` in the "[storage]" section seems to resolve that error for me as well, so I'm assuming it's the same issue. It's too bad that when podman gets updated I feel like I frequently have to fiddle with storage.conf to get things working again.

```
$ podman version
Version:            1.7.0
RemoteAPI Version:  1
Go Version:         go1.13.5
OS/Arch:            linux/amd64
```

**Comment by rhatdan** [comment]
Sorry, I actually think it was me that introduced this bug.  Hopefully the container/storage PR fixes it.

**Comment by mheon** [comment]
I think we can close this; the fix has been merged into c/storage and we just vendored a new version of storage into master. Should be fixed in 1.7.1
